core/src/main/java/com/radixdlt/consensus/vertexstore/VertexStoreAdapter.java

```diff
@@ -150,24 +140,31 @@ public List<ExecutedVertex> getPathFromRoot(HashCode vertexHash) {
 
   public void insertVertexChain(VertexChain vertexChain) {
     final var result = vertexStore.insertVertexChain(vertexChain);
-    result
-        .insertedQcs()
-        .forEach(
-            insertedQc -> {
-              this.highQCUpdateDispatcher.dispatch(
-                  BFTHighQCUpdate.create(insertedQc.vertexStoreState()));
-              insertedQc
-                  .committedUpdate()
-                  .onPresent(
-                      committedUpdate ->
-                          this.bftCommittedDispatcher.dispatch(
-                              new BFTCommittedUpdate(
-                                  committedUpdate.committedVertices(),
-                                  insertedQc.vertexStoreState())));
-            });
+    result.insertedQcs().forEach(this::dispatchPostQcInsertionEvents);
     result.insertUpdates().forEach(bftUpdateDispatcher::dispatch);
   }
 
+  private void dispatchPostQcInsertionEvents(VertexStore.InsertQcResult.Inserted inserted) {
+    // This is a bit of a hack.
+    // We don't want to persist the vertex store state (via PersistentVertexStore)
+    // if the state is already being persisted alongside commit.
```

Comment:
Would be useful if this comment explained why? Is this a performance improvement or does it have correctness implications?

Reply:
Does this fix a race condition in the tests? So it sounds like this is a correctness issue? Why though?

Is it that when we commit a vertex chain, we only want to dispatch the most recent vertex store state? Not the historic vertex store state? Perhaps we should make this even more explicit?

But hmm this doesn't even seem to protect against that though? If A, B and C are inserted, maybe they all have the same `committedUpdate`? It's been too long since I looked at this

Reply:
hmh, I think that was specifically needed around node restarts, e.g. when we manage to persist a vertex store state, but not the commit itself